File: 208. Implement Trie (Prefix Tree)/Solution.java
class TrieNode {
    public TrieNode[] children;
    public boolean isEnd = false;
    public TrieNode() {
        children = new TrieNode[26];
    }
}

class Trie {

    private TrieNode root ;
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        var currentHead = root;
        for (int i = 0; i < word.length(); i++) {
            var idx = word.charAt(i) - 'a';
            if (currentHead.children[idx] == null) {
                currentHead.children[idx] = new TrieNode();
            } 
            currentHead = currentHead.children[idx];   
        }
        currentHead.isEnd = true;
        
    }
    
    public boolean search(String word) {
        var currentHead = root;
        for (int i = 0; i < word.length(); i++) {
            var idx = word.charAt(i) - 'a';
            if (currentHead.children[idx] == null) {
                return false;
            } 
            currentHead = currentHead.children[idx];   
        }
        return currentHead.isEnd;
    }
    
    public boolean startsWith(String word) {
        var currentHead = root;
        for (int i = 0; i < word.length(); i++) {
            var idx = word.charAt(i) - 'a';
            if (currentHead.children[idx] == null) {
                return false;
            } 
            currentHead = currentHead.children[idx];   
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208. Implement Trie (Prefix Tree)/Solution.go
package main

type Node struct {
	children [26]*Node
	isEnd bool
}

type Trie struct {
    head *Node
}


func Constructor() Trie {
    return Trie{head: &Node{}}
}


func (this *Trie) Insert(word string)  {
    currentHead := this.head
	for _, c := range word {
		if currentHead.children[c-'a'] == nil  {
			currentHead.children[c-'a'] = &Node{}
		}
		currentHead = currentHead.children[c-'a']
	}
	currentHead.isEnd = true
}


func (this *Trie) Search(word string) bool {
    currentHead := this.head
	for _, c := range word {
		if currentHead.children[c-'a'] == nil  {
			return false
		}
	}
	return currentHead.isEnd
}


func (this *Trie) StartsWith(prefix string) bool {
    currentHead := this.head
	for _, c := range prefix {
		if currentHead.children[c-'a'] == nil  {
			return false
		}
		currentHead = currentHead.children[c-'a']
	}
	return true
}


/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */